File: src/snapshot.ts
// =============================================================================
// State Snapshot — Capture and hash environment state
//
// Before every agent action, the full state is serialized and hashed.
// The serialization is RFC 8785 (JCS) JSON canonicalization, which produces
// byte-identical output across implementations regardless of insertion order,
// number representation quirks, or unicode normalization differences.
//
// SHA-256 over JCS-canonical bytes is the hash used for both state hashes and
// entry hashes. See spec/SPEC.md §5–6.
// =============================================================================

import { createHash } from "crypto";
import canonicalize from "canonicalize";

/**
 * Compute SHA-256 hex digest of a UTF-8 string.
 */
export function sha256(data: string): string {
  return createHash("sha256").update(data).digest("hex");
}

/**
 * RFC 8785 canonicalization of arbitrary JSON-compatible data.
 *
 * Returns the canonical JSON string. For `undefined` input, returns
 * `undefined` (matches JCS behavior — `undefined` is not JSON).
 */
export function serializeState(state: unknown): string {
  return canonicalize(state) ?? "";
}

/**
 * Capture a full state snapshot: deep-clone the value (for storage) and
 * hash its canonical form (for the ledger).
 */
export function captureSnapshot(
  state: unknown
): { serialized: unknown; hash: string } {
  const serialized = state === undefined ? undefined : JSON.parse(JSON.stringify(state));
  const hash = sha256(serializeState(serialized));
  return { serialized, hash };
}

/**
 * Compute a ledger entry hash from its components (per spec §6.2).
 *
 * Hash = SHA-256(JCS({ sequence, action, stateBefore, stateAfter, parentHash, critic }))
 *
 * JCS sorts keys lexicographically, so the order in which we construct the
 * payload object is irrelevant. Cross-language conformance follows from the
 * canonicalization, not from object construction discipline.
 */
export function computeEntryHash(
  sequence: number,
  action: unknown,
  stateBefore: string,
  stateAfter: string,
  parentHash: string,
  critic?: unknown
): string {
  const payload = { sequence, action, stateBefore, stateAfter, parentHash, critic };
  return sha256(serializeState(payload));
}

/**
 * Verify the integrity of a ledger chain (per spec §7).
 * Returns the index of the first corrupted entry if tampered.
 */
export function verifyChain(
  entries: Array<{
    sequence: number;
    action: unknown;
    stateBefore: string;
    stateAfter: string;
    parentHash: string;
    hash: string;
    critic?: unknown;
  }>
): { valid: boolean; corruptedAt?: number } {
  const GENESIS_HASH = "0".repeat(64);

  for (let i = 0; i < entries.length; i++) {
    const entry = entries[i];

    if (entry.sequence !== i) {
      return { valid: false, corruptedAt: i };
    }

    if (i === 0 && entry.parentHash !== GENESIS_HASH) {
      return { valid: false, corruptedAt: 0 };
    }

    if (i > 0 && entry.parentHash !== entries[i - 1].hash) {
      return { valid: false, corruptedAt: i };
    }

    const expectedHash = computeEntryHash(
      entry.sequence,
      entry.action,
      entry.stateBefore,
      entry.stateAfter,
      entry.parentHash,
      entry.critic
    );

    if (entry.hash !== expectedHash) {
      return { valid: false, corruptedAt: i };
    }
  }

  return { valid: true };
}

File: python/src/map/core/snapshot.py
"""State snapshots, hashing, and chain verification (sync).

Implements SPEC.md §5–§7. Pure CPU — no I/O. The architectural constraint
(DESIGN.md §2) is that I/O lives only in store classes; this module is part
of the pure-CPU core that the v0.2 async retrofit will reuse unchanged.
"""

from __future__ import annotations

import copy
import hashlib
from typing import Any

from .canonical import canonical_bytes, canonical_str

GENESIS_HASH: str = "0" * 64


def sha256_hex(data: bytes | str) -> str:
    """SHA-256 hex digest. Encodes ``str`` to UTF-8 first."""
    if isinstance(data, str):
        data = data.encode("utf-8")
    return hashlib.sha256(data).hexdigest()


def state_hash(state: Any) -> str:
    """SPEC.md §6.1 — SHA-256 of the JCS canonical encoding of state."""
    return sha256_hex(canonical_bytes(state))


def capture_snapshot(state: Any) -> tuple[Any, str]:
    """Deep-clone a state value (for storage) and hash its canonical form.

    Returns ``(serialized_clone, hash_hex)``. ``None`` state is preserved
    rather than deep-cloned.
    """
    serialized = None if state is None else copy.deepcopy(state)
    return serialized, state_hash(serialized)


def compute_entry_hash(
    sequence: int,
    action: Any,
    state_before: str,
    state_after: str,
    parent_hash: str,
    critic: Any | None = None,
) -> str:
    """SPEC.md §6.2 — SHA-256 of the JCS-canonical encoding of the hash payload.

    Payload fields are exactly: sequence, action, stateBefore, stateAfter,
    parentHash, critic. Their order in the dict literal is irrelevant — JCS
    sorts keys lexicographically.
    """
    payload = {
        "sequence": sequence,
        "action": action,
        "stateBefore": state_before,
        "stateAfter": state_after,
        "parentHash": parent_hash,
        "critic": critic,
    }
    return sha256_hex(canonical_bytes(payload))


def verify_chain(entries: list[dict[str, Any]]) -> dict[str, Any]:
    """SPEC.md §7 — verify a ledger.

    Each entry in ``entries`` must be a dict with at least the chained
    fields (sequence, action, stateBefore, stateAfter, parentHash, hash,
    critic).

    Returns ``{"valid": True}`` on success, or
    ``{"valid": False, "corruptedAt": <int>}`` on the first detected
    corruption.
    """
    for i, entry in enumerate(entries):
        if entry.get("sequence") != i:
            return {"valid": False, "corruptedAt": i}

        if i == 0 and entry.get("parentHash") != GENESIS_HASH:
            return {"valid": False, "corruptedAt": 0}

        if i > 0 and entry.get("parentHash") != entries[i - 1].get("hash"):
            return {"valid": False, "corruptedAt": i}

        expected = compute_entry_hash(
            entry["sequence"],
            entry["action"],
            entry["stateBefore"],
            entry["stateAfter"],
            entry["parentHash"],
            entry.get("critic"),
        )

        if entry.get("hash") != expected:
            return {"valid": False, "corruptedAt": i}

    return {"valid": True}


__all__ = [
    "GENESIS_HASH",
    "canonical_bytes",
    "canonical_str",
    "capture_snapshot",
    "compute_entry_hash",
    "sha256_hex",
    "state_hash",
    "verify_chain",
]
